### PR TITLE
Add text generation prediction support for IPUSeq2SeqTrainer

### PR DESCRIPTION
**Pull request by kundaMwiza**
# What does this PR do?

- Feat: support for text generation when using the `IPUSeq2SeqTrainer`. Currently users have to defer to using pipelines in order to run inference workloads
- Fix: the default eval batch size set is `1`, however the docstring specifies `8` 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/graphcore/generation_utils.py [hunk 1]
```diff
@@ -59,14 +59,13 @@ def __init__(self, pipelined_model):
     def forward(self, t, **model_inputs):
         """
         Args:
-            t : (`torch.Tensor(int)`) Tensor with single int representing the current length of the sequence being generated
+            t : (`torch.Tensor(int)`) Tensor with shape [global_batch_size, 1]. All elements are the same int which represents the current length of the sequence being generated
```
Comment (jimypbr) [inline optimum/graphcore/generation_utils.py:62]: Can we not do this whole expand/reshape on device? I'd prefer to just pass a single `t` "timestep". That will be useful in future optimisations.

Comment (kundaMwiza) [reply optimum/graphcore/generation_utils.py]: I accounted for device iterations in the wrong way, the new code is simpler now

File: optimum/graphcore/generation_utils.py [hunk 2]
```diff
@@ -119,6 +118,16 @@ def _prepare_encoder_decoder_kwargs_for_generation(
 
         return model_kwargs
 
+    def _get_cur_token_logits_tensor(self, token_id):
```
Comment (jimypbr) [inline optimum/graphcore/generation_utils.py:121]: Can't we run this on device instead?

File: optimum/graphcore/generation_utils.py [hunk 3]
```diff
@@ -273,12 +282,13 @@ def greedy_search(
 
             # forward pass to get next token
             outputs = self._call_generate(
-                t=torch.tensor(cur_len - 1),
+                t=self._get_cur_token_logits_tensor(cur_len - 1),
```
Comment (jimypbr) [inline optimum/graphcore/generation_utils.py:285]: can't we run `get_cur_token_logits_tensor` on device instead?

Comment (jimypbr) [reply optimum/graphcore/generation_utils.py:285]: I don't think device_iterations is even going to work with text_generation in its current form. Every call to decoder is a single batch. The encoder and decoder use Sharded execution as well (not pipelined parallel).

**Comment by kundaMwiza** [comment]
`python
from optimum.graphcore import pipeline, IPUConfig, IPUSeq2SeqTrainingArguments, IPUSeq2SeqTrainer
from optimum.graphcore import PipelinedT5ForConditionalGeneration
from transformers import AutoModelForSeq2SeqLM
from transformers import AutoTokenizer
from datasets import load_dataset
from transformers import DataCollatorForSeq2Seq
import numpy as np
import evaluate
import poptorch
import torch

if __name__ == "__main__":
    model_checkpoint = "t5-small" 
    source_lang = "en"
    target_lang = "fr"
    lang_map = {"en": "English", "fr": "French", "ro": "Romanian", "ge": "German"}
    max_length = 20 # default max length for en_to_fr
    num_beams = 5
    truncation=True
    padding="max_length"
    micro_batch_size=3
    gradient_accumulation_steps=16
    pod_type="pod4"
    
    books = load_dataset("opus_books", f"{source_lang}-{target_lang}")
    books = books["train"].train_test_split(0.2)
    
    prefix = f"translate {lang_map[source_lang]} {lang_map[target_lang]}: "
    tokenizer = AutoTokenizer.from_pretrained(model_checkpoint, padding=padding, max_length=max_length, truncation=truncation)
    
    # trainer version
    def preprocess_function(examples):
        inputs = [prefix + example[source_lang] for example in examples["translation"]]
        targets = [example[target_lang] for example in examples["translation"]]
        model_inputs = tokenizer(inputs, padding=padding, max_length=max_length, truncation=truncation)
        with tokenizer.as_target_tokenizer():
            targets = tokenizer(targets, padding=padding, max_length=max_length, truncation=truncation)
        # ignore padded labels in the loss
        targets["input_ids"] = [
            [(l if l != tokenizer.pad_token_id else -100) for l in target] for target in targets["input_ids"]
        ]
        model_inputs["labels"] = targets["input_ids"]
        return model_inputs
    
    tokenized_dataset = books.map(preprocess_function, batched=True)

    ipu_config = IPUConfig.from_pretrained("Graphcore/t5-small-ipu", layers_per_ipu=[0, 6, 0, 6])
    model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)
    data_collator = DataCollatorForSeq2Seq(tokenizer=tokenizer, model=model)
    trainer_args = IPUSeq2SeqTrainingArguments(
        "batched-generation",
        evaluation_strategy = "epoch",
        learning_rate=2e-5,
        per_device_train_batch_size=micro_batch_size,
        per_device_eval_batch_size=micro_batch_size,
        gradient_accumulation_steps=gradient_accumulation_steps,
        pod_type=pod_type,
        weight_decay=0.01,
        save_total_limit=3,
        do_train=True,
        do_eval=True,
        num_train_epochs=1,
        dataloader_drop_last=True,
        logging_steps=100,
        push_to_hub=False,
        predict_with_generate=True,
        generation_max_length=max_length,
        generation_num_beams=num_beams
    )
    
    trainer = IPUSeq2SeqTrainer(
        model=model,
        ipu_config=ipu_config,
        data_collator=data_collator,
        args=trainer_args,
        train_dataset=tokenized_dataset["train"],
        eval_dataset=tokenized_dataset["test"],
        compute_metrics=None
    )
    
    # trainer.train()
    output = trainer.predict(tokenized_dataset["test"].select(range(micro_batch_size * ipu_config.inference_device_iterations)))
    trainer_preds, trainer_labels = (output.predictions, output.label_ids)
    
    metric = evaluate.load("sacrebleu")
    
    def postprocess_text(preds, labels):
        preds = [pred.strip() for pred in preds]
        labels = [[label.strip()] for label in labels]
        return preds, labels

    def compute_metrics(eval_preds):
        preds, labels = eval_preds
        if isinstance(preds, tuple):
            preds = preds[0]
        decoded_preds = tokenizer.batch_decode(preds, skip_special_tokens=True)
        labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
        decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)
        decoded_preds, decoded_labels = postprocess_text(decoded_preds, decoded_labels)
        
        result = metric.compute(predictions=decoded_preds, references=decoded_labels)
        result = {"bleu": result["score"]}
        
        prediction_lens = [np.count_nonzero(pred != tokenizer.pad_token_id) for pred in preds]
        result["gen_len"] = np.mean(prediction_lens)
        result = {k: round(v, 4) for k,v in result.items()}
        return result, decoded_preds, decoded_labels
    
    score, decoded_preds, decoded_labels = compute_metrics((trainer_preds, trainer_labels))
    
    # pipeline version
    pipe_data = [[prefix + sample["translation"][source_lang],sample["translation"][target_lang]]  for sample in books["test"].select(range(micro_batch_size * ipu_config.inference_device_iterations))]
    pipe_inputs, pipe_labels = zip(*pipe_data)
    pipe = pipeline("text2text-generation", tokenizer=tokenizer, model=model_checkpoint, ipu_config=ipu_config)
    outputs = pipe(list(pipe_inputs), max_input_length=max_length, max_length=max_length, num_beams=num_beams, truncation=truncation)
    
    print("-"*50)
    for i in range (len(decoded_preds)):
        print(f"trainer prediction: {decoded_preds[i]} \npipeline prediction: {outputs[i]['generated_text']}")
        print(f"label: {pipe_labels[i]}")
        print("-"*50)
    
    
`

File: optimum/graphcore/trainer_seq2seq.py [hunk 4]
```diff
@@ -31,9 +31,37 @@ class IPUSeq2SeqTrainer(IPUTrainer):
...
+        """
+        Detach inference model from IPUs
+        Override for text generation specific behaviour
+        """
+        # for text generation:
+        # IPUGenerationMixin::_call_generate() handles the compilation of the
+        # poplar executor, It adds the compiled models
+        # as poptorch_decoder and poptorch_encoder to the model object, so the actual
+        # inference_models we need to destroy are poptorch_decoder and poptorch_encoder
+        if hasattr(self.inference_model, "poptorch_decoder"):
```
[the remark this answers is not in the document]
Comment (jimypbr) [reply optimum/graphcore/trainer_seq2seq.py:55]: there is already a method `.detachFromDevice` in `IPUGenerationMixin`

[the remark this answers is not in the document]
Comment (kundaMwiza) [reply optimum/graphcore/trainer_seq2seq.py:55]: Yep, but until the sdk 3.2 branch is merged with 'main', 'super()' handles the case for tied weights models. If not, it's a breaking change on 'main' until then?

[the remark this answers is not in the document]
Comment (kundaMwiza) [reply optimum/graphcore/trainer_seq2seq.py:55]: I can remove this and land the PR after the sdk3.2 branch has been merged with main?

[the remark this answers is not in the document]
Comment (jimypbr) [reply optimum/graphcore/trainer_seq2seq.py:55]: Yes that sounds good. merging sdk3.2 in the next few day.
